File: 0101-0200/0154.Find_Minimum_in_Rotated_Sorted_Array_II.cpp
//0ms




class Solution {
public:
    //----------Optimal Approach----------//
    int findMin(vector<int>& nums) {
        int n = nums.size();
        int min_ele = nums[0];
        int low = 0, high = n - 1;

        if (n == 1)
            return nums[0];

        while (low <= high) {
            int rem_ele = high - low + 1;
            int mid = (low + high) / 2;

            // Case when all three are equal and we can't decide the sorted half
            if (rem_ele > 3 && nums[low] == nums[mid] && nums[mid] == nums[high]) {
                low++;
                high--;
                continue;
            }

            // Left half is sorted
            if (nums[low] <= nums[mid]) {
                min_ele = min(min_ele, nums[low]);
                low = mid + 1;
            }
            // Right half is sorted
            else {
                min_ele = min(min_ele, nums[mid]);
                high = mid - 1;
            }
        }

        return min_ele;
    }
};





//0ms






class Solution {
public:
    int findMin(vector<int>& nums) {
        int n = nums.size();
        int ans = nums[0];

        for (int i = 1; i < n; i++) {
            if (nums[i] < nums[i - 1]) {
                ans = nums[i];
            }
        }
        return ans;
    }
};

File: 0101-0200/0154.Find_Minimum_in_Rotated_Sorted_Array_II.rs
//0ms



impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let (mut left, mut right): (usize, usize) = (0, nums.len()-1);
        
        if nums[left] < nums[right] {
            return nums[left];
        }
        
        while left < right {
            let mid = left + (right-left)/2;
            
            if nums[mid] > nums[right] {
                left = mid+1;
            } else if nums[mid] < nums[right] {
                right = mid;
            } else {
                right-=1;  //avoid duplication
            }
        }
        
        nums[right]
    }
}





//0ms





impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let mut min = i32::MAX;
        for n in nums {
            min = min.min(n);
        }
        min
    }
}
